Allow using modules as subtypes of protocols

ilevkivskyi opened the pull request:
Fixes #5018 
Fixes #5439
Fixes #10850 

The implementation is simple but not the most beautiful one. I simply add a new slot to the `Instance` class that represents content of the module. This new attribute is short lived (it is not serialized, and not even stored on variables etc., because we erase it in `copy_modified()`). We don't need to store it, because all the information we need is already available in `MypyFile` node. We just need the new attribute to communicate between the checker and `subtypes.py`.

Other possible alternatives like introducing new dedicated `ModuleType`, or passing the symbol tables to `subtypes.py` both look way to complicated. Another argument in favor of this new slot is it could be useful for other things, like `hasattr()` support and ad hoc callable attributes (btw I am already working on the former).

Note there is one important limitation: since we don't store the module information, we can't support module objects stored in nested positions, like `self.mods = (foo, bar)` and then `accepts_protocol(self.mods[0])`. We only support variables (name expressions) and direct instance, class, or module attributes (see tests). I think this will cover 99% of possible use-cases.

ilevkivskyi commented:
Hm, those few errors are clearly wrong, and I know where they are coming from. Will try to fix now.

sobolevn commented on test-data/unit/check-protocols.test:
```diff
@@ -3546,3 +3546,142 @@ S = TypeVar("S")
+import bad_runner
+from typing import Callable, Protocol
+
+class Runner(Protocol):
```
What if some protocol requires `ClassVar`?

```python
class My(Protocol):
    A: ClassVar[int]
```

Can this be satisfied by a module (I think not)? Do we need a test for this?

sobolevn commented on test-data/unit/check-protocols.test:
```diff
@@ -3546,3 +3546,142 @@ S = TypeVar("S")
+                 # N:     Got: \
+                 # N:         def __init__(arg: str) -> Run
+
+[file runner.py]
```
I think it makes sence to test something like:

```python
from typing_extensions import Protocol
class My(Protocol):
   a: int  # is settable attr

def func(arg: My): pass

import some_module
func(some_module)  # error
```

And:

```python
# some_module.py
from typing_extensions import Final

a: Final = 1  # is module-level constant
```

Looks like we should correctly raise an error here.

ilevkivskyi commented:
Yes, I think both should be errors. I will add tests later today (and update implementation if needed).

ilevkivskyi commented:
@sobolevn OK, I added test and fixed code. Btw I discovered that we didn't do the final check for normal instances either, so I fixed that as well (I actually even found an issue about this on the tracker).

sobolevn reviewed (approved):
🎉

gvanrossum commented:
You rock! Also thank you to the reviewers.

ilevkivskyi commented:
Got some free time on my paternity leave :-)

gvanrossum commented:
> Got some free time on my paternity leave :-)

Oh congratulations!

ilevkivskyi commented:
Thank you!
